File: packages/virage-cli/src/output.ts
import { ansi } from "./ansi.js";

const DIVIDER_WIDTH = 60;

export interface Out {
  error(msg: string): void;
  warn(msg: string): void;
  success(msg: string): void;
  info(msg: string): void;
  dim(msg: string): void;
  verbose(msg: string): void;
  debug(msg: string): void;
  section(label: string): void;
  divider(char?: string, width?: number, color?: string): void;
}

export function createOut(verbosity: number): Out {
  return {
    error: (msg) => console.error(`${ansi.boldRed}✕ ${msg}${ansi.reset}`),
    warn: (msg) => console.warn(`${ansi.yellow}⚠ ${msg}${ansi.reset}`),
    success: (msg) => console.log(`${ansi.green}✓ ${msg}${ansi.reset}`),
    info: (msg) => console.log(msg),
    dim: (msg) => console.log(`${ansi.dim}${msg}${ansi.reset}`),
    verbose: (msg) => {
      if (verbosity >= 1) console.log(`${ansi.dim}  ${msg}${ansi.reset}`);
    },
    debug: (msg) => {
      if (verbosity >= 2)
        console.log(`${ansi.dimGray}  [debug] ${msg}${ansi.reset}`);
    },
    section: (label) => {
      const line = `${ansi.dim}${"─".repeat(DIVIDER_WIDTH)}${ansi.reset}`;
      console.log(`\n${line}`);
      console.log(`${ansi.bold}${ansi.cyan} ${label}${ansi.reset}`);
      console.log(line);
    },
    divider: (char = "─", width = DIVIDER_WIDTH, color = ansi.dim) =>
      console.log(`${color}${char.repeat(width)}${ansi.reset}`),
  };
}

// Zero-verbosity fallback for callers that haven't been updated yet.
export const out = createOut(0);

File: crates/virage-engine/src/output.rs
use console::style;

const DIVIDER_WIDTH: usize = 60;

#[derive(Clone, Copy, PartialEq, Debug)]
pub enum OutputFormat {
    Human,
    Json,
    Quiet,
}

pub struct Out {
    pub verbosity: u8,
    pub format: OutputFormat,
}

impl Out {
    pub fn new(verbosity: u8, format: OutputFormat) -> Self {
        Self { verbosity, format }
    }

    // ── Human-only output (stderr) ──────────────────────────────

    pub fn section(&self, label: &str) {
        if self.format != OutputFormat::Human {
            return;
        }
        let line = style("─".repeat(DIVIDER_WIDTH)).dim().to_string();
        eprintln!("\n{line}");
        eprintln!("{}", style(format!(" {label}")).bold().cyan());
        eprintln!("{line}");
    }

    pub fn success(&self, msg: &str) {
        if self.format != OutputFormat::Human {
            return;
        }
        eprintln!("{} {msg}", style("✓").green());
    }

    pub fn info(&self, msg: &str) {
        if self.format != OutputFormat::Human {
            return;
        }
        eprintln!("{msg}");
    }

    pub fn dim(&self, msg: &str) {
        if self.format != OutputFormat::Human {
            return;
        }
        eprintln!("{}", style(msg).dim());
    }

    // ── Warning — human+quiet to stderr ────────────────────────

    pub fn warn(&self, msg: &str) {
        if self.format == OutputFormat::Json {
            return;
        }
        eprintln!("{} {msg}", style("⚠").yellow());
    }

    // ── Error — human/quiet to stderr; json to stdout ──────────

    pub fn error(&self, msg: &str) {
        match self.format {
            OutputFormat::Json => {
                println!(
                    "{{\"error\":{}}}",
                    serde_json::to_string(msg).unwrap_or_default()
                );
            }
            _ => {
                eprintln!("{} {msg}", style("✕").red().bold());
            }
        }
    }

    pub fn error_hint(&self, msg: &str, hint: &str) {
        match self.format {
            OutputFormat::Json => {
                println!(
                    "{{\"error\":{},\"hint\":{}}}",
                    serde_json::to_string(msg).unwrap_or_default(),
                    serde_json::to_string(hint).unwrap_or_default(),
                );
            }
            _ => {
                eprintln!("{} {msg}", style("✕").red().bold());
                eprintln!("   Hint: {hint}");
            }
        }
    }

    // ── Data output (stdout always) ─────────────────────────────

    pub fn data_json(&self, value: &serde_json::Value) {
        if self.format == OutputFormat::Json {
            println!(
                "{}",
                serde_json::to_string_pretty(value).unwrap_or_default()
            );
        }
    }

    pub fn data_line(&self, msg: &str) {
        if self.format != OutputFormat::Json {
            println!("{msg}");
        }
    }

    // ── Verbosity levels (stderr) ───────────────────────────────

    pub fn verbose(&self, msg: &str) {
        if self.format == OutputFormat::Human && self.verbosity >= 1 {
            eprintln!("{}", style(format!("  {msg}")).dim());
        }
    }

    pub fn debug_msg(&self, msg: &str) {
        if self.format == OutputFormat::Human && self.verbosity >= 3 {
            eprintln!("{}", style(format!("  [debug] {msg}")).color256(240).dim());
        }
    }

    pub fn trace_msg(&self, msg: &str) {
        if self.format == OutputFormat::Human && self.verbosity >= 5 {
            eprintln!("{}", style(format!("  [trace] {msg}")).dim());
        }
    }
}
